Add documentation

nezia1 opened the pull request:
This PR adds support for documentation through the means of [ndg v2](https://github.com/feel-co/ndg/tree/v2). This PR is a draft as I have not been able to figure out *yet* how to generate a proper optionsJSON, as some attributes seem to be missing for a proper evaluation: 

```console
       error: attribute 'users' missing
       at /nix/store/z7hkn7gh9xjx2m4kaxz3a5rwzpyr122x-source/modules/nixos/default.nix:27:25:
           26|             user = config.users.users.${name}.name;
           27|             directory = config.users.users.${name}.home;
             |                         ^
           28|             clobberFiles = cfg.clobberByDefault;
```

Unsure how I could remedy to this. @NotAShelf could you please take a look at this and tell me what I'm doing wrong?

Lunarnovaa reviewed (commented):
This is looking great, is it possible to setup the docs generator into the pre-commit hooks or as a github action?

nezia1 commented:
> This is looking great, is it possible to setup the docs generator into the pre-commit hooks or as a github action?

It is, but considering this is a local thing, it would be better to set this up as a GitHub actions pipeline only.

Lunarnovaa reviewed (commented):
Rather than having ` ``${config.directory}/.config/file` it would probably be better to have something like `~/.config/file`  

Once xdg is impl upstream, we could do something like `$XDG_CONFIG_HOME/file`.

Alternatively, we could just do the `$XDG_CONFIG_HOME` from the getgo.

nezia1 commented:
> Rather than having ` ``${config.directory}/.config/file` it would probably be better to have something like `~/.config/file`  
> 
> 
> 
> Once xdg is impl upstream, we could do something like `$XDG_CONFIG_HOME/file`.
> 
> 
> 
> Alternatively, we could just do the `$XDG_CONFIG_HOME` from the getgo.


I think doing it from the getgo is fine, but it will be slightly misleading. The former is probably better for now since it's essentially a single search and replace throughout out our modules.

nezia1 commented:
> Current issues related to the docs generation:
> 
>     1. Paths are not being transformed yet, and just show a path to the nix store of the machine that generated documentation. This is something that can be fixed if we don't use `ndg-builder` and use the tool by itself,  but let's wait until this is fixed
Fixed with latest version of ndg 

>     2. `_module.args` is not being filtered out from our options list
This is still happening, unsure if it will be implemented upstream.  But the rest of it works just fine, and does link to our main branch. I would like to make it so that it links to the proper revision, which I'm unsure of how to do.

nezia1 commented:
Things done according to what we discussed:

1. Filtered out `_module.args` out of the options
2. Implemented CI (we just need to set it up accordingly in the repo settings after we merge this)
3. Updated paths to mention `~` instead of `${config.directory}` until Hjem implements xdg user dirs

This PR should be ready now. Please let me know if you want to add more to the online user manual, I'd be more than happy to document specific parts that might be more nebulous.

NotAShelf commented:
I can provide a review sometime tomorrow.

nezia1 commented:
So, I had to change a few things to fix merge conflicts, and accomodate our new `rumLib` separation. This caused issues in `docs/package.nix`, mainly with the scrubber. Right now, from my limited understanding of it, I believe that the scrubber needs to evaluate alongside the list of modules.

However, this is going to be an issue as our collection of modules require `rumLib`. The current solution I found (which is not great), is to do this:

```nix
      (evalModules {
        modules =
          (lib.modules.importApply ../modules/hjem.nix {inherit lib rumLib;}).imports
          ++ [
```

Basically, we grab the list of modules from imports, and then concatenate that to our scrubber, which will work since it will have access to `rumLib`via `importApply`. I hate this though, and I would like to find a better solution. Maybe @NotAShelf has an idea, but from my limited Nix knowledge, I can't figure this one out.

NotAShelf commented on .github/workflows/manual.yml:
```diff
@@ -0,0 +1,55 @@
+      - uses: actions/checkout@v4.1.7
+      - uses: DeterminateSystems/nix-installer-action@main
+      - run: |
+          nix build .#docs -Lv
+          cp -r result public
```
`nix build -Lv .#docs -o `public` unless you need to write to the directory. `-o` stands for `--out-link`

NotAShelf commented on .github/workflows/manual.yml:
```diff
@@ -0,0 +1,55 @@
+      - uses: peaceiris/actions-gh-pages@v4
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: ./public
```
Why not publish from `result` then?

NotAShelf commented on flake.nix:
```diff
@@ -3,13 +3,22 @@
+      url = "github:feel-co/hjem";
+      inputs.nixpkgs.follows = "nixpkgs";
+    };
+    ndg = {
+      url = "github:feel-co/ndg/v2";
+    };
```
Since `ndg` has no cache (yet) you can follow `nixpkgs` until I get around to creating one.

NotAShelf replied:
This only happens to you, there is no breakage in recent nixpkgs.

Lunarnovaa replied:
Is this bug still occurring? We really should try to have their nixpkgs input follows ours.

NotAShelf commented on flake.nix:
```diff
@@ -27,7 +36,14 @@
+    packages = forAllSystems (system: let
+      pkgs = nixpkgs.legacyPackages.${system};
+    in {
+      docs = import ./docs/package.nix {
```
Packages are `callPackage`'d, not `import`ed.

NotAShelf commented on modules/collection/misc/gtk.nix:
```diff
@@ -59,7 +59,7 @@ in {
         default = "";
         description = ''
-          CSS to be written to '${config.directory}/.config/gtk-3.0/gtk.css'.
+          CSS to be written to `~/.config/gtk-3.0/gtk.css`.
```
```suggestion
          CSS to be written to {file}`~/.config/gtk-3.0/gtk.css`.
```

ndg supports nixpkgs flavors, and stylizes those MySt-kind roles.

nezia1 replied:
Should I be able to nest those MySt roles? I tried doing `{file}`{env}`HOME`/.config/alacritty/alacritty.toml`.`, but that did not render correctly with `ndg`. It shows `{env} HOME` verbatim.

nezia1 replied:
Ended up just doing `{file}`$HOME` until we get `$XDG_CONFIG_HOME` into Hjem, that should be fine. Let me know what you think.

NotAShelf replied:
No, nesting is not supported since we cannot parse it.

NotAShelf commented on modules/collection/misc/gtk.nix:
```diff
@@ -68,7 +68,7 @@ in {
         default = "";
         description = ''
-          CSS to be written to '${config.directory}/.config/gtk-4.0/gtk.css'.
+          CSS to be written to `~/.config/gtk-4.0/gtk.css`.
```
Same as above.

NotAShelf commented on modules/collection/programs/alacritty.nix:
```diff
@@ -34,8 +34,8 @@ in {
         The configuration converted into TOML and written to
-        `${config.directory}/.config/alacritty/alacritty.toml`.
-        Please reference https://alacritty.org/config-alacritty.html
+        `~/.config/alacritty/alacritty.toml`.
+        Please reference [https://alacritty.org/config-alacritty.html](https://alacritty.org/config-alacritty.html)
```
You might be able to use automatic links here, e.g., `[](link)` but I don't undeerstand why the link body and the title are the same here?

nezia1 replied:
That does not render *yet* with `ndg`. I did this because I just want to render the link with the same text,  but I did not know you could omit the brackets part. It completely hides the link though, with ndg's parsing.

NotAShelf replied:
That is a bug, please report all missing nixpkgs flavors.

NotAShelf commented on modules/collection/programs/foot.nix:
```diff
@@ -40,10 +40,10 @@ in {
       };
       description = ''
-        Settings are written as an INI file to ${config.directory}/.config/foot/foot.ini.
+        Settings are written as an INI file to ~/.config/foot/foot.ini.
```
```suggestion
        Settings are written as an INI file to {file}`~/.config/foot/foot.ini`.
```

NotAShelf commented on modules/collection/programs/fuzzel.nix:
```diff
@@ -27,7 +27,7 @@ in {
       };
       description = ''
-        Is written to `${config.directory}/fuzzel/fuzzel.ini`.
+        Is written to `~/fuzzel/fuzzel.ini`.
```
```suggestion
        Is written to {file}`~/fuzzel/fuzzel.ini`.
```

NotAShelf commented on modules/collection/programs/fuzzel.nix:
```diff
@@ -27,7 +27,7 @@ in {
-        Is written to `${config.directory}/fuzzel/fuzzel.ini`.
+        Is written to `~/fuzzel/fuzzel.ini`.
 
         Consult [man 5 fuzzel.ini](https://www.mankier.com/5/fuzzel.ini).
```
Nixpkgs flavored markdown (and ndg too, naturally) supports automatic manpage links through the [manpage role](https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#roles)

nezia1 replied:
The issue is that fuzzel and others do not have a manpage under [manpage-urls.json](https://github.com/NixOS/nixpkgs/blob/master/doc/manpage-urls.json). Unless I should just pass a custom JSON file completely?

NotAShelf replied:
Using a custom manpage list would be ideal.

nezia1 replied:
> Using a custom manpage list would be ideal.

Done, however it does not seem to work. Could you please take a look to see if it's a ndg issue? E.g., this:
```nix
# programs/tofi.nix
...
      description = ''
        The configuration converted into "key = value" and written to
        {file}`$HOME/.config/tofi/config`.
        Please reference {manpage}`tofi(5)`,
        or see an example at [https://github.com/philj56/tofi/blob/master/doc/config](https://github.com/philj56/tofi/blob/master/doc/config).
      '';
```
Renders to this:
```console
The configuration converted into “key = value” and written to $HOME/.config/tofi/config. Please reference tofi(5)tofi(5), or see an example at https://github.com/philj56/tofi/blob/master/doc/config.
```

NotAShelf reviewed (dismissed):
I'm leaving a few comments on areas I think are brittle. You are free to not address them as HJR is not my project, but I think it's nicer to use the nixpkgs flavored markdown features for documentation consistent with Nixpkgs (...and Home-Manager), while also testing ndg's support for those features :)

Please see https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#syntax


P.S. I'm not sure why the path is hardcoded as `~`, can the user not pick a different directory to manage with HJR? Not all shells expand `~` either, so it's a bit weird. Perhaps use something like

```
{env}`HOME`
```

Lunarnovaa commented on docs/manpage-urls.json:
```diff
@@ -0,0 +1,5 @@
+{
+"fuzzel.ini(5)": "https://man.archlinux.org/man/fuzzel.ini.5",
+"tofi(5)": "https://github.com/philj56/tofi/blob/master/doc/tofi.5.md",
+"ncmpcpp": "https://manpages.debian.org/testing/ncmpcpp/ncmpcpp.1.en.html"
```
Should it be like this or am I mistaken?

```suggestion
"ncmpcpp(1)": "https://manpages.debian.org/testing/ncmpcpp/ncmpcpp.1.en.html"
```

btw, this file is indented weirdly, is that just JSON or perhaps should the contents of the attrset be indented

Lunarnovaa commented on docs/package.nix:
```diff
@@ -0,0 +1,117 @@
+      };
+  };
+
+  html = pkgs.runCommandLocal "hjem-rum-docs" {} ''
+    mkdir -p $out/share/man/man5
+    ${getExe ndg} options \
+    --input-dir ${./manual} \
+    --module-options ${hjemRumDocs.optionsJSON}/share/doc/nixos/options.json \
+    --manpage-urls ${./manpage-urls.json} \
+    --options-depth 3 \
+    --revision "https://github.com/snugnug/hjem-rum/tree/main" \
+    --jobs $NIX_BUILD_CORES \
+    --generate-search true \
+    --title "Hjem Rum" \
+    --output-dir $out/share/man/man5/hjem-rum.5
+  '';
+in
+  html
```
NDGv2 now has a module for docs generation, yes? Any thoughts on using that for readability?

nezia1 replied:
It didn't use to work, probably because I did some funky stuff when I first started working it. But it does now :)

Lunarnovaa commented on flake.nix:
```diff
@@ -3,13 +3,22 @@
+    ndg,
     ...
-  }: let
+  } @ inputs: let
```
`@ inputs` isn't necessary if we're not using it. Especially in this flake since we've got like 3 inputs, one of them being nixpkgs.

```suggestion
  }: let
```

nezia1 replied:
We are using it (it's passed into `docs/package.nix` as a special argument)

Lunarnovaa replied:
In that case, just use ndg as inputs.ndg rather than passing it through the attrset.

Lunarnovaa commented on modules/collection/programs/alacritty.nix:
```diff
@@ -34,8 +34,8 @@ in {
-        `${config.directory}/.config/alacritty/alacritty.toml`.
-        Please reference https://alacritty.org/config-alacritty.html
+        {file}`$HOME/.config/alacritty/alacritty.toml`.
+        Please reference [https://alacritty.org/config-alacritty.html](https://alacritty.org/config-alacritty.html)
```
```suggestion
        Please reference [Alacritty's documentation](https://alacritty.org/config-alacritty.html)
```

Lunarnovaa commented on modules/collection/programs/foot.nix:
```diff
@@ -40,10 +40,10 @@ in {
 
-        Refer to https://codeberg.org/dnkl/foot/src/branch/master/foot.ini for
-        all available options.
+        Refer to [https://codeberg.org/dnkl/foot/src/branch/master/foot.ini](https://github.com/aome510/spotify-player/blob/master/docs/config.md#themes)
+        for all available options.
```
```suggestion
        Refer to {manpage}`foot.ini(5)` or the [upstream template](https://codeberg.org/dnkl/foot/src/branch/master/foot.ini)
        for all available options.
```

Also add the manpage to `docs/manpage-urls.json`.

Lunarnovaa commented on modules/collection/programs/ghostty.nix:
```diff
@@ -46,8 +46,9 @@ in {
-        The configuration converted to INI and written to `${config.directory}/.config/ghostty/config`.
-        Please reference https://ghostty.org/docs/config/reference for config options.
+        The configuration converted to INI and written to {file}`$HOME/.config/ghostty/config`.
+        Please reference [https://ghostty.org/docs/config/reference](https://ghostty.org/docs/config/reference)
+        for config options.
```
```suggestion
        Please reference [Ghostty's documentation](https://ghostty.org/docs/config/reference)
        for config options.
```

Lunarnovaa commented on modules/collection/programs/ghostty.nix:
```diff
@@ -83,7 +84,8 @@ in {
       description = ''
         An attribute set of themes, with the key as the theme name.
-        Please reference https://ghostty.org/docs/features/theme for config options.
+        Please reference [https://ghostty.org/docs/features/theme](https://ghostty.org/docs/features/theme)
+        for config options.
```
```suggestion
        Please reference [Ghostty's documentation](https://ghostty.org/docs/features/theme)
        for config options.
```

Lunarnovaa commented on modules/collection/programs/helix.nix:
```diff
@@ -41,10 +41,10 @@ in {
-        and https://docs.helix-editor.com/remapping.html for config
-        options.
+        {file}`$HOME/.config/helix/config.toml`.
+        Please reference [https://docs.helix-editor.com/editor.html](https://docs.helix-editor.com/editor.html)
+        and [https://docs.helix-editor.com/remapping.html](https://docs.helix-editor.com/remapping.html) for
+        config options.
```
```suggestion
        Please reference Helix's [editor configuration](https://docs.helix-editor.com/editor.html)
        and [remapping](https://docs.helix-editor.com/remapping.html) documentation for
        config options.
```

Lunarnovaa commented on modules/collection/programs/helix.nix:
```diff
@@ -56,8 +56,8 @@ in {
-        `${config.directory}/.config/helix/languages.toml`.
-        Please reference https://docs.helix-editor.com/languages.html
+        {file}`$HOME/.config/helix/languages.toml`.
+        Please reference [https://docs.helix-editor.com/languages.html](https://docs.helix-editor.com/languages.html)
```
```suggestion
        Please reference Helix's [language configuration](https://docs.helix-editor.com/languages.html) documentation
```

Lunarnovaa commented on modules/collection/programs/helix.nix:
```diff
@@ -77,8 +77,8 @@ in {
-        `${config.directory}/.config/helix/themes/`.
-        Please reference https://docs.helix-editor.com/themes.html
+        {file}`$HOME/.config/helix/themes/`.
+        Please reference [https://docs.helix-editor.com/themes.html](https://docs.helix-editor.com/themes.html)
```
```suggestion
        Please reference Helix's [theming](https://docs.helix-editor.com/themes.html) documentation
```

Lunarnovaa commented on modules/collection/programs/hyprland.nix:
```diff
@@ -33,7 +33,8 @@ in {
         Hyprland configuration written in Nix. Entries with the same key
         should be written as lists. Variables' and colors' names should be
-        quoted. See <https://wiki.hyprland.org> for more examples.
+        quoted. See [https://wiki.hyprland.org](https://wiki.hyprland.org)
+        for more examples.
```
```suggestion
        quoted. See [Hyprland's documentation](https://wiki.hyprland.org)
        for more examples.
```

Lunarnovaa commented on modules/collection/programs/keepassxc.nix:
```diff
@@ -34,11 +34,11 @@ in {
-        Please consult https://keepassxc.org/docs/KeePassXC_UserGuide, but also
-        a configuration you create by toggling options through the GUI, as it
-        doesn't seem they are documented.
+        Please consult [https://keepassxc.org/docs/KeePassXC_UserGuide](https://keepassxc.org/docs/KeePassXC_UserGuide),
+        but also a configuration you create by toggling options through the GUI, as it doesn't seem
+        they are documented.
```
```suggestion
        Please consult [KeePassXC's User Guide](https://keepassxc.org/docs/KeePassXC_UserGuide),
        but also a configuration you create by toggling options through the GUI, as it doesn't seem
        they are documented.
```

Lunarnovaa commented on modules/collection/programs/lsd.nix:
```diff
@@ -26,8 +26,9 @@ in {
-        Configuration written to `${config.directory}/.config/lsd/config.yaml`, defining lsd settings.
-        Please reference https://github.com/lsd-rs/lsd#config-file-content to configure it accordingly.
+        Configuration written to {file}`$HOME/.config/lsd/config.yaml`, defining lsd settings.
+        Please reference [https://github.com/lsd-rs/lsd#config-file-content](https://github.com/lsd-rs/lsd#config-file-content)
+        to configure it accordingly.
```
```suggestion
        Please reference [lsd's example config](https://github.com/lsd-rs/lsd#config-file-content)
        to configure it accordingly.
```

Lunarnovaa commented on modules/collection/programs/lsd.nix:
```diff
@@ -42,9 +43,9 @@ in {
-        lsd.
-        Please reference https://github.com/lsd-rs/lsd#icon-theme to configure it accordingly.
+        Configuration written to {file}`$HOME/.config/lsd/icons.yaml`, defining the icons used by lsd.
+        Please reference [https://github.com/lsd-rs/lsd#icon-theme](https://github.com/lsd-rs/lsd#icon-theme)
+        to configure it accordingly.
```
```suggestion
        Please reference [lsd's icon theme example](https://github.com/lsd-rs/lsd#icon-theme)
        to configure it accordingly.
```

Lunarnovaa reviewed (changes requested):
Looks great. I think this can be merged once NDG v2 is merged to main, if not sooner. I left some comments on each of the files with some suggestions. Kinda got tired and realized I should've just made my commits straight to your branch and stopped halfway through. You get the idea though, lol.

NotAShelf commented:
All things considered NDG v2 is good to go as is. Maybe, if the HJR team is interested, I'll add relevant CLI flags to

1. disable internal syntax highlighting
2. specify multiple script files (so that, e.g., highlight.js can be loaded as an alternative)

nezia1 commented:
First batch of suggestions taken care of. The markdown links have been updated to have a proper text

nezia1 commented:
`ndg-builder` has also been added to simplify the codebase

nezia1 commented:
> All things considered NDG v2 is good to go as is. Maybe, if the HJR team is interested, I'll add relevant CLI flags to
> 
>     1. disable internal syntax highlighting
> 
>     2. specify multiple script files (so that, e.g., highlight.js can be loaded as an alternative)

This is not needed for now. However, could it be possible to add a parameter for `manpage-urls.json`? As there is no current way to pass it into the builder

nezia1 commented:
I decided against using the builder, as per a discussion with @NotAShelf (it's too opinionated and it's not possible to use `transformOpts` with it, which we need to render links instead of store paths). I essentially went back to the old version I had, with a couple of fixes (better usage of `nativeBuildInputs`, use of `internal = true` to not render `_module.args` in the documentation etc.). Let me know if anything is wrong or odd :)

NotAShelf commented on docs/CONTRIBUTING.md:
```diff
@@ -146,6 +146,10 @@ configured by the user.
 
 ### Writing Options<a name="writing-options"></a>
 
+> \[!IMPORTANT\] When writing options for any Nix module, do NOT make any option
```
If I'm not mistaken the admoniation (`\[!IMPORTANT\]`) needs to be its own line.

nezia1 replied:
Yup, this was a mistake from a previous commit that I made, I don't know what went wrong with the formatting. Also, I don't think using `README.md` and `CONTRIBUTING.md`as pages are a good idea in hindsight, so I just took it out.

NotAShelf reviewed (commented):
Everything LGTM. I think you'd benefit from using masked links like

```markdown
[mask]: https://link.com

Here is a [mask]ed link for you
```

for clarity. I'm quite sure I support those out of the box *without* literalMD, but please do let me know if it doesn't work.

nezia1 commented:
> Everything LGTM. I think you'd benefit from using masked links like
> 
> ```
> [mask]: https://link.com
> 
> Here is a [mask]ed link for you
> ```
> 
> for clarity. I'm quite sure I support those out of the box _without_ literalMD, but please do let me know if it doesn't work.

This should be taken care of in #54 (I want to do documentation sweeps gradually)

Lunarnovaa commented on modules/lib/generators/ncmpcpp.nix:
```diff
@@ -39,7 +39,9 @@ in {
           else "no";
       }
-      .${typeOf value};
+      .${
+        typeOf value
+      };
```
What's the reasoning for this? Personally I feel like this is less clear. But it's not a big deal.

nezia1 replied:
This is from the new version of `alejandra` i believe. It did format like that as I ran `nix fmt`, so I thought I'd include it here as a final commit

Lunarnovaa replied:
Alright, that's totally fine then.

Lunarnovaa reviewed (changes requested):
Almost good. Also fix the package examples missing the literalExpression.

Lunarnovaa commented on docs/package.nix:
```diff
@@ -0,0 +1,119 @@
+
+      transformOptions = opt:
+        opt
+        // {
```
nevermind, im a dumbass

nezia1 replied:
The reason why we're doing this is so that we can have `opt` keep its original value, and add `declarations` to the attrset. Doing it the way you suggested would override `opt` completely

nezia1 commented on docs/package.nix:
```diff
@@ -0,0 +1,119 @@
+                # avoid having `_module.args` in the documentation
+                {
+                  options = {
+                    _module.args = pkgs.lib.mkOption {
```
This was a mistake. Using it rn

nezia1 commented on modules/collection/misc/gtk.nix:
```diff
@@ -17,18 +16,21 @@
-        })
-        pkgs.bibata-cursors
-      ];
+      example = literalExpression ''
+        [
+          (pkgs.catppuccin-papirus-folders.override {
+            accent = "rosewater";
+            flavor = "mocha";
+          })
+          pkgs.bibata-cursors
+        ];
+      '';
```
It would be, we just need to have an additive `--stylesheet` argument on ndg

nezia1 commented on modules/collection/programs/ncmpcpp.nix:
```diff
@@ -19,16 +19,15 @@ in {
-        package = pkgs.ncmpccpp.override {
+      '';
+      example = ''
+          pkgs.ncmpcpp.override {
```
This is how `alejandra` indents it apparently

Lunarnovaa commented:
Please rebase onto the newest commit, since the tests PR some conflicting changes that will need to be resolved.
